lib/RexLint.php

```diff
@@ -88,4 +111,51 @@ private static function linterBinPath(): string
 
         return $path;
     }
+
+    /**
+     * @return array<string, list<array{line: int, message: string}>>
+     */
+    private static function validateAddOnsPackageYml():array {
+        $packageSchema = rex_path::core('schemas/package.json');
+
+        $errorsPerFile = [];
+        foreach(\rex_package::getAvailablePackages() as $package) {
+            $packageYml = $package->getPath('package.yml');
+            if (!is_file($packageYml)) {
+                continue;
+            }
+
+            $jsonData = rex_file::getConfig($packageYml);
+            foreach(self::validateJsonSchema($jsonData, $packageSchema) as $error) {
+                if (!array_key_exists($packageYml, $errorsPerFile)) {
+                    $errorsPerFile[$packageYml] = [];
+                }
+
+                $errorsPerFile[$packageYml][] = $error;
+            }
+        }
+
+        return $errorsPerFile;
+    }
+
+    /**
+     * @param array<mixed> $json
+     *
+     * @return list<array{line: int, message: string}>
+     */
+    private static function validateJsonSchema(array $json, string $schemaPath): array {
+        $validator = new \JsonSchema\Validator();
+        $validator->validate($json, (object) ['$ref' => 'file://'.$schemaPath], \JsonSchema\Constraints\Constraint::CHECK_MODE_TYPE_CAST);
+
+        $errors = [];
+        if (!$validator->isValid()) {
+            foreach ($validator->getErrors() as $error) {
+                $errors[] = [
+                    'line' => 0,
+                    'message' => ($error['property'] ? $error['property'].' : ' : '').$error['message']
+                ];
+            }
+        }
+        return $errors;
+    }
```

Comment:
@seldaek I saw you are a maintainer of justinrainbow/json-schema which I am using here.


may I ask you if you have an idea why a json structure like (dumped using symfony dumper).

```
array:7 [▼
    "package" => "install"
    "version" => "2.11.1"
    "author" => "Gregor Harlan"
    "supportpage" => "https://github.com/redaxo/redaxo"
    "page" => array:7 [▼
        "title" => "translate:title"
        "perm" => "admin[]"
        "block" => "system"
        "prio" => 70
        "pjax" => true
        "icon" => "rex-icon rex-icon-install"
        "subpages" => array:2 [▼
            "packages" => array:2 [▼
                "title" => "translate:subpage_packages"
                "subpages" => array:3 [▼
                    "update" => array:1 [▼
                        "title" => "translate:subpage_update"
                    ]
                    "add" => array:1 [▼
                        "title" => "translate:subpage_add"
                    ]
                    "upload" => array:1 [▼
                        "title" => "translate:subpage_upload"
                    ]
                ]
            ]
            "settings" => array:1 [▼
                "title" => "translate:subpage_settings"
            ]
        ]
    ]
    "requires" => array:2 [▼
        "php" => array:2 [▼
            "version" => ">=8.1"
            "extensions" => array:1 [▼
                0 => "zlib"
            ]
        ]
        "redaxo" => "^5.15.0"
    ]
    "console_commands" => array:3 [▼
        "install:download" => "rex_command_install_download"
        "install:list" => "rex_command_install_list"
        "install:update" => "rex_command_install_update"
    ]
]
```

errors with

<img width="702" alt="grafik" src="https://user-images.githubusercontent.com/120441/229303435-c0d2c262-c5da-425a-9930-d5bbca8d7039.png">



when validating against this schema:
https://github.com/redaxo/redaxo/blob/b88ed5891e159773b8324e33f9602e1ad19d850e/redaxo/src/core/schemas/package.json


-----

using the very same schema directly within phpstorm does not report any errors for said json structure.

I have spent several hours already on it :-/

Reply:
In the main-page schema try anyOf instead of allOf maybe? Sorry not sure I'm not very familiar with refs and merging like this..

Reply:
great catch, thank you.